flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/GlobalViewDeclarativeSlotPoolBridge.java

```diff
@@ -0,0 +1,261 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+
+/**
+ * {@link SlotPool} implementation which could use the {@link GlobalViewDeclarativeSlotPoolBridge}
+ * to allocate slots in a global view. Note: It's only used for streaming mode now.
+ */
+public class GlobalViewDeclarativeSlotPoolBridge extends DeclarativeSlotPoolBridge
+        implements TimeoutListener<GlobalViewDeclarativeSlotPoolBridge> {
+
+    public static final Logger LOG =
+            LoggerFactory.getLogger(GlobalViewDeclarativeSlotPoolBridge.class);
+    private final Map<SlotRequestId, Boolean> increasedResourceRequirements;
+
+    private final TimerService<GlobalViewDeclarativeSlotPoolBridge> timerService;
+
+    private final @Nonnull Set<PhysicalSlot> receivedNewSlots;
+
+    private final @Nonnull Map<SlotRequestId, PhysicalSlot> preFulfilledFromAvailableSlots;
+    private final Time slotRequestMaxInterval;
+
+    public GlobalViewDeclarativeSlotPoolBridge(
+            JobID jobId,
+            DeclarativeSlotPoolFactory declarativeSlotPoolFactory,
+            Clock clock,
+            Time rpcTimeout,
+            Time idleSlotTimeout,
+            Time batchSlotTimeout,
+            Time slotRequestMaxInterval,
+            RequestSlotMatchingStrategy requestSlotMatchingStrategy) {
+        super(
+                jobId,
+                declarativeSlotPoolFactory,
+                clock,
+                rpcTimeout,
+                idleSlotTimeout,
+                batchSlotTimeout,
+                requestSlotMatchingStrategy);
+        this.slotRequestMaxInterval = Preconditions.checkNotNull(slotRequestMaxInterval);
+        this.receivedNewSlots = new HashSet<>();
+        this.preFulfilledFromAvailableSlots = new HashMap<>();
+        this.increasedResourceRequirements = new HashMap<>();
+        this.timerService =
+                new DefaultTimerService<>(
+                        new ScheduledThreadPoolExecutor(1),
+                        slotRequestMaxInterval.toMilliseconds());
+    }
+
+    @Override
+    protected void internalRequestNewAllocatedSlot(PendingRequest pendingRequest) {
+        pendingRequests.put(pendingRequest.getSlotRequestId(), pendingRequest);
+        increasedResourceRequirements.put(pendingRequest.getSlotRequestId(), false);
+
+        timerService.registerTimeout(
+                this, slotRequestMaxInterval.getSize(), slotRequestMaxInterval.getUnit());
+    }
+
+    @Override
+    void newSlotsAreAvailable(Collection<? extends PhysicalSlot> newSlots) {
+        receivedNewSlots.addAll(newSlots);
+        if (newSlots.isEmpty() && receivedNewSlots.isEmpty()) {
+            // TODO: Do the matching logic only for available slots.
+        } else {
+            // TODO: Do the matching logic for new slots and available slots.
```

Comment:
I didn't finish the review, so I didn't understand why it has some TODOs here?

As I understand, this commit just introduce 2 timeout mechanism, right?

1. slotRequestMaxInterval
2. Whether start assign slots after all slots are ready

Reply:
> I didn't finish the review, so I didn't understand why it has some TODOs here?

Yes, you're right.~

> After these 2 conditions are true, we call the original assign logic(It means we just introduce timeout mechanism, and don't break the assignment process), right?

The implementation of the 2 TODOs is completed in the corresponding test class, because the the implementation in the test class is not the final implementation at the `2 TODOs` point.